File: backup/cyclesignal_flutter/lib/logic/cycle.dart
import 'package:intl/intl.dart';
import '../models/cycle_entry.dart';
import '../theme/app_theme.dart';
import 'package:flutter/material.dart';

// --- Date helpers (date-only) ---
DateTime parseIso(String d) => DateTime.parse('${d}T00:00:00');
String toIsoDate(DateTime d) => DateFormat('yyyy-MM-dd').format(d);
DateTime addDays(DateTime d, int n) => DateTime(d.year, d.month, d.day + n);
int diffDays(DateTime a, DateTime b) => a.difference(b).inDays;
String monthLabel(DateTime d) => DateFormat('MMM yyyy').format(d);
String prettyDate(String iso) => DateFormat('EEE d MMM yyyy').format(parseIso(iso));
DateTime todayDate() {
  final n = DateTime.now();
  return DateTime(n.year, n.month, n.day);
}

class Cycle {
  final int cycleNumber;
  final String startDate;
  final String label;
  final int length;
  final bool isIrregular;
  final List<CycleEntry> entries;
  Cycle({
    required this.cycleNumber,
    required this.startDate,
    required this.label,
    required this.length,
    required this.isIrregular,
    required this.entries,
  });
}

List<Cycle> buildCycles(List<CycleEntry> entries) {
  final byNumber = <int, List<CycleEntry>>{};
  for (final e in entries) {
    byNumber.putIfAbsent(e.cycleNumber, () => []).add(e);
  }
  final numbers = byNumber.keys.toList()..sort();
  final starts = <int, DateTime>{};
  for (final n in numbers) {
    final dates = byNumber[n]!.map((e) => parseIso(e.date)).toList();
    dates.sort();
    starts[n] = dates.first;
  }
  final today = todayDate();

  final out = <Cycle>[];
  for (var i = 0; i < numbers.length; i++) {
    final n = numbers[i];
    final start = starts[n]!;
    final cycleEntries = byNumber[n]!..sort((a, b) => a.cycleDay.compareTo(b.cycleDay));
    int length;
    if (i < numbers.length - 1) {
      length = diffDays(starts[numbers[i + 1]]!, start);
    } else {
      final maxDay = cycleEntries.map((e) => e.cycleDay).fold<int>(1, (p, c) => c > p ? c : p);
      final elapsed = diffDays(today, start) + 1;
      length = maxDay > elapsed ? maxDay : elapsed;
    }
    out.add(Cycle(
      cycleNumber: n,
      startDate: toIsoDate(start),
      label: 'Cycle $n — ${monthLabel(start)}',
      length: length,
      isIrregular: length > 35,
      entries: cycleEntries,
    ));
  }
  return out;
}

enum CellCategory { none, bleeding, moodLow, moodMid, moodHigh, physical }

class Cell {
  final int cycleNumber;
  final int cycleDay;
  final CellCategory category;
  final Color color;
  final bool hasFunctionalImpact;
  final bool safetyFlagged;
  final CycleEntry? entry;
  final bool premenstrual;
  Cell({
    required this.cycleNumber,
    required this.cycleDay,
    required this.category,
    required this.color,
    required this.hasFunctionalImpact,
    required this.safetyFlagged,
    required this.entry,
    required this.premenstrual,
  });
}

({CellCategory category, Color color}) _categorise(CycleEntry? e) {
  if (e == null) return (category: CellCategory.none, color: AppColors.cellNoData);
  final b = e.bleeding;
  if (b.started || (b.flow != null && b.flow != 'none')) {
    return (category: CellCategory.bleeding, color: AppColors.cellBleeding);
  }
  if (e.mood?.present == true) {
    final s = e.mood!.severity;
    if (s >= 4) return (category: CellCategory.moodHigh, color: AppColors.cellMoodHigh);
    if (s == 3) return (category: CellCategory.moodMid, color: AppColors.cellMoodMid);
    return (category: CellCategory.moodLow, color: AppColors.cellMoodLow);
  }
  final p = e.physical;
  final physical = (p.pain?.present ?? false) ||
      (p.fatigue?.present ?? false) ||
      p.acne ||
      p.hairGrowth ||
      p.bloating;
  if (physical) return (category: CellCategory.physical, color: AppColors.cellPhysical);
  return (category: CellCategory.none, color: AppColors.cellNoData);
}

class Heatmap {
  final int maxDay;
  final List<List<Cell>> rows;
  Heatmap({required this.maxDay, required this.rows});
}

Heatmap buildHeatmap(List<Cycle> cycles) {
  if (cycles.isEmpty) return Heatmap(maxDay: 1, rows: const []);
  var maxLen = cycles.map((c) => c.length).fold<int>(1, (p, c) => c > p ? c : p);
  final maxDay = maxLen > 65 ? 65 : maxLen;
  final rows = cycles.map((cycle) {
    final byDay = {for (final e in cycle.entries) e.cycleDay: e};
    final cells = <Cell>[];
    for (var day = 1; day <= maxDay; day++) {
      final entry = byDay[day];
      final c = _categorise(entry);
      final premenstrual = day > cycle.length - 10 && day <= cycle.length;
      cells.add(Cell(
        cycleNumber: cycle.cycleNumber,
        cycleDay: day,
        category: c.category,
        color: c.color,
        hasFunctionalImpact: entry?.functionalImpact.present ?? false,
        safetyFlagged: entry?.safetyFlagged ?? false,
        entry: entry,
        premenstrual: premenstrual,
      ));
    }
    return cells;
  }).toList();
  return Heatmap(maxDay: maxDay, rows: rows);
}

List<String> describeEntry(CycleEntry e) {
  final out = <String>[];
  if (e.mood?.present == true) {
    final types = e.mood!.types.isEmpty ? 'low' : e.mood!.types.join(', ');
    out.add('Mood: $types (severity ${e.mood!.severity})');
  }
  if (e.physical.pain?.present == true) {
    out.add('Pain (severity ${e.physical.pain!.severity})');
  }
  if (e.physical.fatigue?.present == true) {
    out.add('Fatigue (severity ${e.physical.fatigue!.severity})');
  }
  if (e.physical.sleep?.disrupted == true) {
    out.add('Disrupted sleep (${e.physical.sleep!.nightsAffected} nights)');
  }
  if (e.physical.acne) out.add('Acne');
  if (e.physical.hairGrowth) out.add('Hair growth');
  if (e.physical.bloating) out.add('Bloating');
  if (e.bleeding.started || (e.bleeding.flow != null && e.bleeding.flow != 'none')) {
    out.add('Bleeding${e.bleeding.flow != null ? ' (${e.bleeding.flow})' : ''}');
  }
  if (e.functionalImpact.present) {
    out.add('Functional impact: ${e.functionalImpact.types.join(', ')}');
  }
  if (e.medications.isNotEmpty) out.add('Medications: ${e.medications.join(', ')}');
  return out.isEmpty ? ['No symptoms logged'] : out;
}

File: backup/cyclesignal_rn/app/src/lib/cycle.ts
import { CycleEntry } from '../types';
import { colors } from '../theme/colors';

// --- Date helpers (UTC-safe, date-only) ---
export function parseISO(d: string): Date {
  return new Date(d + 'T00:00:00');
}
export function toISODate(d: Date): string {
  return d.toISOString().slice(0, 10);
}
export function addDays(d: Date, n: number): Date {
  const r = new Date(d);
  r.setDate(r.getDate() + n);
  return r;
}
export function diffDays(a: Date, b: Date): number {
  return Math.round((a.getTime() - b.getTime()) / 86400000);
}
export function monthLabel(d: Date): string {
  return d.toLocaleDateString('en-GB', { month: 'short', year: 'numeric' });
}
export function prettyDate(iso: string): string {
  return parseISO(iso).toLocaleDateString('en-GB', {
    weekday: 'short',
    day: 'numeric',
    month: 'short',
    year: 'numeric',
  });
}

export type Cycle = {
  cycleNumber: number;
  startDate: string;
  label: string;
  length: number;
  isIrregular: boolean;
  entries: CycleEntry[];
};

// Group entries into cycles and derive length from the gap between successive cycle starts.
export function buildCycles(entries: CycleEntry[]): Cycle[] {
  const byNumber = new Map<number, CycleEntry[]>();
  for (const e of entries) {
    const arr = byNumber.get(e.cycleNumber) ?? [];
    arr.push(e);
    byNumber.set(e.cycleNumber, arr);
  }
  const numbers = [...byNumber.keys()].sort((a, b) => a - b);
  const starts = new Map<number, Date>();
  for (const n of numbers) {
    const dates = byNumber.get(n)!.map((e) => parseISO(e.date).getTime());
    starts.set(n, new Date(Math.min(...dates)));
  }
  const today = new Date(new Date().toISOString().slice(0, 10) + 'T00:00:00');

  return numbers.map((n, i) => {
    const start = starts.get(n)!;
    const cycleEntries = byNumber.get(n)!.slice().sort((a, b) => a.cycleDay - b.cycleDay);
    let length: number;
    if (i < numbers.length - 1) {
      length = diffDays(starts.get(numbers[i + 1])!, start);
    } else {
      // Current cycle: longest known so far, but at least days elapsed.
      const maxDay = Math.max(...cycleEntries.map((e) => e.cycleDay));
      length = Math.max(maxDay, diffDays(today, start) + 1);
    }
    return {
      cycleNumber: n,
      startDate: toISODate(start),
      label: `Cycle ${n} — ${monthLabel(start)}`,
      length,
      isIrregular: length > 35,
      entries: cycleEntries,
    };
  });
}

export type CellCategory = 'none' | 'bleeding' | 'moodLow' | 'moodMid' | 'moodHigh' | 'physical';

export type Cell = {
  cycleNumber: number;
  cycleDay: number;
  category: CellCategory;
  color: string;
  hasFunctionalImpact: boolean;
  safetyFlagged: boolean;
  entry: CycleEntry | null;
  premenstrual: boolean; // within days -10..-1 of cycle length
};

// Dominant-symptom colour for a single cell — PRD Feature 2 / Section 0.7.
function categorise(entry: CycleEntry | null): { category: CellCategory; color: string } {
  if (!entry) return { category: 'none', color: colors.cellNoData };
  if (entry.bleeding.started || (entry.bleeding.flow && entry.bleeding.flow !== 'none')) {
    return { category: 'bleeding', color: colors.cellBleeding };
  }
  if (entry.mood?.present) {
    const s = entry.mood.severity;
    if (s >= 4) return { category: 'moodHigh', color: colors.cellMoodHigh };
    if (s === 3) return { category: 'moodMid', color: colors.cellMoodMid };
    return { category: 'moodLow', color: colors.cellMoodLow };
  }
  const physical =
    entry.physical.pain?.present ||
    entry.physical.fatigue?.present ||
    entry.physical.acne ||
    entry.physical.hairGrowth ||
    entry.physical.bloating;
  if (physical) return { category: 'physical', color: colors.cellPhysical };
  return { category: 'none', color: colors.cellNoData };
}

// Build the full heatmap grid: one row per cycle, columns up to the longest cycle.
export function buildHeatmap(cycles: Cycle[]): { maxDay: number; rows: Cell[][] } {
  const maxDay = Math.min(65, Math.max(1, ...cycles.map((c) => c.length)));
  const rows = cycles.map((cycle) => {
    const byDay = new Map<number, CycleEntry>();
    for (const e of cycle.entries) byDay.set(e.cycleDay, e);
    const cells: Cell[] = [];
    for (let day = 1; day <= maxDay; day++) {
      const entry = byDay.get(day) ?? null;
      const { category, color } = categorise(entry);
      const premenstrual = day > cycle.length - 10 && day <= cycle.length;
      cells.push({
        cycleNumber: cycle.cycleNumber,
        cycleDay: day,
        category,
        color,
        hasFunctionalImpact: !!entry?.functionalImpact.present,
        safetyFlagged: !!entry?.safetyFlagged,
        entry,
        premenstrual,
      });
    }
    return cells;
  });
  return { maxDay, rows };
}

// Human-readable symptom list for an entry (tooltip + history).
export function describeEntry(entry: CycleEntry): string[] {
  const out: string[] = [];
  if (entry.mood?.present) {
    out.push(`Mood: ${entry.mood.types.join(', ') || 'low'} (severity ${entry.mood.severity})`);
  }
  if (entry.physical.pain?.present) out.push(`Pain (severity ${entry.physical.pain.severity})`);
  if (entry.physical.fatigue?.present)
    out.push(`Fatigue (severity ${entry.physical.fatigue.severity})`);
  if (entry.physical.sleep?.disrupted)
    out.push(`Disrupted sleep (${entry.physical.sleep.nightsAffected} nights)`);
  if (entry.physical.acne) out.push('Acne');
  if (entry.physical.hairGrowth) out.push('Hair growth');
  if (entry.physical.bloating) out.push('Bloating');
  if (entry.bleeding.started || (entry.bleeding.flow && entry.bleeding.flow !== 'none'))
    out.push(`Bleeding${entry.bleeding.flow ? ` (${entry.bleeding.flow})` : ''}`);
  if (entry.functionalImpact.present)
    out.push(`Functional impact: ${entry.functionalImpact.types.join(', ')}`);
  if (entry.medications.length) out.push(`Medications: ${entry.medications.join(', ')}`);
  return out.length ? out : ['No symptoms logged'];
}
